text/0000-dockerfiles.md

```diff
@@ -0,0 +1,218 @@
+The `base_image` arg allows the Dockerfile to reference the original base image.
+The `build_id` arg allows the Dockerfile to invalidate the cache after a certain layer and must be defaulted to `0`. The executor of the Dockerfile will provide the `build_id` as a UUID (this eliminates the need to track this variable). 
+When the `$build_id` arg is referenced in a `RUN` instruction, all subsequent layers will be rebuilt on the next build (as the value will change).
+
+Build args specified in `build.toml` are provided to `build.Dockerfile` or `Dockerfile` (when applied to the build-time base image).
+Build args specified in `launch.toml` are provided to `run.Dockerfile` or `Dockerfile` (when applied to the runtime base image).
+
+A runtime base image may indicate that it preserves ABI compatibility by adding the label `io.buildpacks.rebasable=true`. In the case of builder-specified Dockerfiles, `io.buildpacks.rebasable=false` is set automatically on the base image before a runtime Dockerfile is applied and must be explicitly set to `true` if desired. If multiple Dockerfiles are applied, all must set `io.buildpacks.rebasable=true` for the final value to be `true`. Rebasing an app without this label set to `true` requires passing a new `--force` flag to `pack rebase`.
+
+#### Example: App-specified Dockerfile Extension
+
+This example extension would allow an app to provide runtime and build-time base image extensions as "run.Dockerfile" and "build.Dockerfile."
+The app developer can decide whether the extensions are rebasable.
+
+##### `/cnb/ext/com.example.appext/bin/build`
+```
+#!/bin/sh
+[ -f build.Dockerfile ] && cp build.Dockerfile "$1/"
+[ -f run.Dockerfile ] && cp run.Dockerfile "$1/"
+```
+
+#### Example: RPM Dockerfile Extension (app-based)
+
+This example extension would allow a builder to install RPMs for each language runtime, based on the app directory.
+
+Note: The Dockerfiles referenced must disable rebasing, and build times will be slower compared to buildpack-provided runtimes.
+
+##### `/cnb/ext/com.example.rpmext/bin/build`
+```
+#!/bin/sh
+[ -f Gemfile.lock ] && cp "$CNB_BUILDPACK_DIR/Dockerfile-ruby" "$1/Dockerfile"
+[ -f package.json ] && cp "$CNB_BUILDPACK_DIR/Dockerfile-node" "$1/Dockerfile"
+```
+
+
+### Dockerfiles for Base Images
+
+The same Dockerfile format may be used to create new base images or modify existing base images outside of the app build process (e.g., before creating a builder). Any specified labels override existing values.
+
+Dockerfiles that are used to create a base image must create a `/cnb/image/genpkgs` executable that outputs a [CycloneDX](https://cyclonedx.org)-formatted list of packages in the image with PURL IDs when invoked. This executable is executed after all Dockerfiles are applied, and the output replaces the label `io.buildpacks.sbom`. This label doubles as a Software Bill-of-Materials for the base image. In the future, this label will serve as a starting point for the application SBoM.
```

Comment:
```suggestion
Base images may contain a layer with SBoM files as detailed in https://github.com/buildpacks/rfcs/pull/186. After Dockerfiles are applied, the final step in the extension process will be to run a `genpkgs` executable (supplied by the platform, see further details that follow) in the context of the extended image that will generate an updated SBoM, taking as input any files that were already present. The diffID of the layer containing the output from `genkpgs` will replace the `io.buildpacks.base.sbom` label on the extended run image. 

When assembling layers for the extended run image, the extender will exclude the `genpkgs` executable path, so that `genpkgs` will not be present in the extended run image.

The project will ship a reference implementation of `genpkgs` in the [lifecycle image](https://hub.docker.com/r/buildpacksio/lifecycle), a minimal image containing the lifecycle binaries. Platforms when running the extender may choose to override the provided `genpkgs` with their own implementation. We propose that the reference implementation use [syft](https://github.com/anchore/syft).
```

Reply:
@samj1912 @sclevine updated with my understanding from yesterday's OH...

Reply:
> Can we clarify the interplay between genpkgs and any sbom files output by hooks? It's kind of confusing that both mechanisms are allowed. When would you use one vs the other?

I'm still curious about this...